app/upload.go

```diff
@@ -19,6 +19,67 @@ import (
 
 const minFirstPartSize = 5 * 1024 * 1024 // 5MB
 
+func (a *App) runPluginsHook(info *model.FileInfo, file io.Reader) *model.AppError {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil
+	}
+
+	// using a pipe to avoid loading the whole file content in memory.
+	r, w := io.Pipe()
+	errChan := make(chan *model.AppError, 1)
+	go func() {
+		defer w.Close()
+		defer close(errChan)
+		pluginContext := a.PluginContext()
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			newInfo, rejStr := hooks.FileWillBeUploaded(pluginContext, info, file, w)
+			if rejStr != "" {
+				errChan <- model.NewAppError("runPluginsHook", "app.upload.run_plugins_hook.rejected", nil, rejStr, http.StatusBadRequest)
+				return false
+			}
+			if newInfo != nil {
+				info = newInfo
+			}
+			return true
+		}, plugin.FileWillBeUploadedId)
+	}()
+
+	tmpPath := info.Path + ".tmp"
+	written, err := a.WriteFile(r, tmpPath)
+	if err != nil {
+		if fileErr := a.RemoveFile(tmpPath); fileErr != nil {
+			mlog.Error("Failed to remove file", mlog.Err(fileErr))
+		}
+		return err
```

Comment:
There's a bit of a problem here. If we return from here, then we will never be reading from `errChan` so `RunMultiPluginHook` will have a goroutine leak.

Reply:
My understanding is that we don't need to be reading from it for the goroutine to exit since it's using a buffered channel.

Reply:
It's a buffered channel of size 1. But the `RunMultiPluginHook` will be called for each plugin. So it will block whenever we have more than 1 plugin. The buffer needs to be equal to the number of plugins.

Reply:
That's definitely something I failed to consider. A case where more than one plugin has the `FileWillBeUploaded` hook registered.